File: app/src/main/java/com/esorokin/lantector/presentation/presenter/detect/DetectLanguagePresenter.java
package com.esorokin.lantector.presentation.presenter.detect;

import com.arellomobile.mvp.InjectViewState;
import com.esorokin.lantector.di.DependencyManager;
import com.esorokin.lantector.model.data.DetectedLanguageText;
import com.esorokin.lantector.model.interactor.detect.DetectLanguageInteractor;
import com.esorokin.lantector.model.network.exception.NoInternetException;
import com.esorokin.lantector.presentation.error.ErrorProcessor;
import com.esorokin.lantector.presentation.presenter.BasePresenter;
import com.esorokin.lantector.presentation.presenter.EventConsumerTransformer;
import com.esorokin.lantector.presentation.view.detect.DetectLanguageView;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;

@InjectViewState
public class DetectLanguagePresenter extends BasePresenter<DetectLanguageView> {
	@Inject
	DetectLanguageInteractor interactor;

	@Inject
	ErrorProcessor errorProcessor;

	public DetectLanguagePresenter() {
		DependencyManager.getAppComponent().inject(this);

		autoDispose(interactor.detectingEmitter()
				.observeOn(AndroidSchedulers.mainThread())
				.compose(EventConsumerTransformer.<DetectedLanguageText>builder()
						.showLoading(() -> getViewState().showDetectingProgress())
						.hideLoading(() -> getViewState().hideDetectingProgress())
						.showError(this::handleDetectingError)
						.hideError(() -> getViewState().hideError())
						.receiveData(detectedLanguageText -> getViewState().showDetectingResult(detectedLanguageText.getLanguage()))
						.build())
				.subscribe());
	}

	@Override
	public void onDestroy() {
		super.onDestroy();
		interactor.onDestroy();
	}

	private void handleDetectingError(Throwable throwable) {
		if (throwable instanceof NoInternetException) {
			getViewState().noInternetConnection();
		} else {
			getViewState().showError(errorProcessor.processError(throwable));
		}
	}

	public void userClickStartDetect(String textToDetect) {
		interactor.requestDetectLanguage(textToDetect);
	}

	public void userClickRetryDetect(String textToDetect) {
		interactor.requestDetectLanguage(textToDetect);
	}

	public void userHideDetectingResult() {
		getViewState().hideDetectingResult();
	}

	public void userHideError() {
		getViewState().hideError();
	}
}

File: app/src/main/java/com/esorokin/lantector/presentation/presenter/detect/DetectLanguagePresenter.kt
package com.esorokin.lantector.presentation.presenter.detect

import com.arellomobile.mvp.InjectViewState
import com.esorokin.lantector.di.DependencyManager
import com.esorokin.lantector.model.AppErrorType
import com.esorokin.lantector.model.interactor.detect.DetectLanguageInteractor
import com.esorokin.lantector.presentation.error.ErrorProcessor
import com.esorokin.lantector.presentation.presenter.BasePresenter
import com.esorokin.lantector.presentation.view.detect.DetectLanguageView
import com.esorokin.lantector.utils.ext.autoDispose
import com.esorokin.lantector.utils.ext.consumeBy
import com.esorokin.lantector.utils.ext.toAppException
import javax.inject.Inject

@InjectViewState
class DetectLanguagePresenter : BasePresenter<DetectLanguageView>() {
    @Inject
    internal lateinit var interactor: DetectLanguageInteractor

    @Inject
    internal lateinit var errorProcessor: ErrorProcessor

    init {
        DependencyManager.appComponent.inject(this)

        interactor.detectingEmitter()
                .consumeBy(
                        showLoading = { viewState.showDetectingProgress() },
                        hideLoading = { viewState.hideDetectingProgress() },
                        showError = handleDetectingError(),
                        hideError = { viewState.hideError() },
                        receiveData = { viewState.showDetectingResult(it.language) }
                )
                .autoDispose(this)

        interactor.autoDispose(this)
    }

    private fun handleDetectingError(): (Throwable) -> Unit = {
        val appException = it.toAppException()
        when (appException.appErrorType) {
            is AppErrorType.NoInternet -> viewState.noInternetConnection()
            else -> viewState.showError(errorProcessor.processError(it))
        }
    }

    fun userClickStartDetect(textToDetect: String) {
        interactor.requestDetectLanguage(textToDetect)
    }

    fun userClickRetryDetect(textToDetect: String) {
        interactor.requestDetectLanguage(textToDetect)
    }

    fun userHideDetectingResult() {
        viewState.hideDetectingResult()
    }

    fun userHideError() {
        viewState.hideError()
    }
}
